source/game/system/map/MapdataCheckPoint.cc

```diff
@@ -0,0 +1,258 @@
+/// checkpoint line and the next; otherwise, false
+/// @addr{Inlined in 0x80510C74}
+bool MapdataCheckPoint::checkCheckpointCompletion(const LinkedCheckpoint &next,
+        const EGG::Vector2f &p0, const EGG::Vector2f &p1, float *checkpointCompletion) const {
+    f32 d1 = m_dir.dot(p1);
+    f32 d2 = -(next.checkpoint->m_dir.dot(p0));
+    // This is where the divide by zero thing happens @todo
+    f32 checkpointCompletion_ = d1 / (d1 + d2);
+    *checkpointCompletion = checkpointCompletion_;
+    return (checkpointCompletion_ >= 0.0f && checkpointCompletion_ <= 1.0f);
+}
+
+/// @brief Calls both @ref checkSector and @ref checkCheckpointCompletion; updates @param
+/// checkpointCompletion
+/// @addr{0x80510C74}
+MapdataCheckPoint::SectorOccupancy MapdataCheckPoint::checkSectorAndCheckpointCompletion_(
+        const LinkedCheckpoint &next, const EGG::Vector2f &p0, const EGG::Vector2f &p1,
+        float *checkpointCompletion) const {
+    if (!checkSector(next, p0, p1)) {
+        return OutsideSector;
+    }
+
+    return checkCheckpointCompletion(next, p0, p1, checkpointCompletion) ?
+            InsideSector :
+            OutsideSector_BetweenSides;
+}
+
+/// @addr{0x80512064}
+f32 MapdataCheckPointAccessor::calculateMeanTotalDistanceRecursive(u16 ckptId) {
+    f32 sumDist = 0.0f;
+    MapdataCheckPoint *ckpt = get(ckptId);
+    u16 n = ckpt->nextCount();
+    for (size_t i = 0; i < n; i++) {
+        const LinkedCheckpoint &linked = ckpt->nextLinked(i);
+        sumDist += linked.distance;
+        ckpt = ckpt->nextPoint(i);
+        if (m_finishLineCheckpointId == ckpt->id()) {
+            continue;
+        }
+        sumDist += calculateMeanTotalDistanceRecursive(ckpt->id());
+    }
+
+    return sumDist / n;
+}
+
+s8 MapdataCheckPointAccessor::lastKcpType() const {
+    return m_lastKcpType;
+}
+
+/// @addr{80512370}
+f32 MapdataCheckPointAccessor::calculateMeanTotalDistance() {
+    return calculateMeanTotalDistanceRecursive(m_finishLineCheckpointId);
+}
+
+/// @brief find finish line and last key checkpoint indexes; also initCheckpointLinks for all
+/// checkpoints
+/// @addr{Inlined in 0x80515244} fake. not real. it's not in the base game. in the base
+/// game, it's inlined into `init()`. i, kooshnoo, split it out because i wanted to.
+// I'm not sure if this should be split; it feels like a separate function,
+//  but idk if i can somehow prove nintendo wrote it as a separate function,
+//  and i don't know if we care what nintendo extracted to a separate function and what they
+//  inlined. would the PR reviewer kindly please advise, thank you
+void MapdataCheckPointAccessor::findFinishAndLastKcp() {
+    s8 lastKcpType = -1;
+    s16 finishLineCheckpointId = -1;
+    for (size_t ckptId = 0; ckptId < size(); ckptId++) {
+        MapdataCheckPoint *checkpoint = get(ckptId);
+        checkpoint = get(ckptId);
+        if (checkpoint->isFinishLine()) {
```

Comment:
Same here, can you help break up the flow of your code by adding a newline before each of your if-statements, and before/after the for loop?